### PR TITLE
Add support for spaces to the Octopus Deploy plugin

### PR DESCRIPTION
**Pull request by jmezach**
## Hey, I just made a Pull Request!

This continues the work from #17115. Unfortunately I wasn't able to push changes to the original repo, hence the new PR. The Octopus Deploy now supports different spaces by adding a space ID and a slash to the annotation on entities. This is especially useful for people who use Octopus Deploy Spaces feature.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [-] Tests for new functionality and regression tests for bug fixes
- [-] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by jmezach** [comment]
@freben I think this can be reviewed.

File: plugins/octopus-deploy/src/components/EntityPageOctopusDeploy/EntityPageOctopusDeploy.tsx [hunk 1]
```diff
@@ -15,18 +15,23 @@
...
+} from '../../utils/getAnnotationFromEntity';
 import React from 'react';
 import { ReleaseTable } from '../ReleaseTable';
 
 export const EntityPageOctopusDeploy = (props: { defaultLimit?: number }) => {
   const { entity } = useEntity();
 
-  const projectId = getAnnotationFromEntity(entity);
+  const projectId = getProjectIdAnnotationFromEntity(entity);
+  const spaceId = getSpaceIdAnnotationFromEntity(entity);
```
Comment (freben) [inline plugins/octopus-deploy/src/components/EntityPageOctopusDeploy/EntityPageOctopusDeploy.tsx:29]: yeah as mentioned in another comment, a single `const projectReference = getAnnotationFromEntity(entity)` that hides the underlying implementation sounds better here

File: .changeset/red-cheetahs-invite.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-octopus-deploy': minor
+---
+
+Added support for octopus deploy spaces.
```
Comment (freben) [inline .changeset/red-cheetahs-invite.md:5]: Might want to mention in just a short sentence what annotation changes you made, helping users adapt accordingly. Also mention that the arguments to some exported things are now a single object instead of separate args.

**Review by freben — COMMENTED** [review]
OK just a few small questions

File: plugins/octopus-deploy/src/constants.ts [hunk 3]
```diff
@@ -16,3 +16,5 @@
 
 /** @public */
 export const OCTOPUS_DEPLOY_PROJECT_ID_ANNOTATION = 'octopus.com/project-id';
+/** @public */
+export const OCTOPUS_DEPLOY_SPACE_ID_ANNOTATION = 'octopus.com/space-id';
```
[the remark this answers is not in the document]
Comment (jmezach) [reply plugins/octopus-deploy/src/constants.ts:20]: Yeah, I was a bit on the fence about this one. I guess we could use [spaceid/]projectid as they appear as <space-id>/projects/<project-id> in the URL. I'll see if I can make that work.

**Comment by jmezach** [comment]
@freben I've processed your comments and I think it is all good now. Somehow the tests are failing though, but I don't think that it has anything to do with my changes, so I'm not sure what's going on.

**Comment by jmezach** [comment]
@freben Any thoughts on these CI failures?

**Comment by freben** [comment]
Yeah those are fixed in master, so you can rebase on that

**Comment by jmezach** [comment]
Ah, I see. I've rebased this branch and all looks good now :). Thanks for the help!

**Comment by jmezach** [comment]
@freben Can this be merged now?

File: plugins/octopus-deploy/README.md [hunk 4]
```diff
@@ -42,9 +42,9 @@ const cicdContent = (
 )
 ```
 
-3. Add `octopus.com/project-id` annotation in catalog descriptor file
+3. Add `octopus.com/project-id` and optionally an `octopus.com/space-id` annotation in the catalog descriptor file.
```
Comment (freben) [inline plugins/octopus-deploy/README.md:45]: This is no longer the case, right

File: plugins/octopus-deploy/package.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@backstage/plugin-octopus-deploy",
-  "version": "0.1.2-next.0",
+  "version": "0.1.1-next.1",
```
Comment (freben) [inline plugins/octopus-deploy/package.json:3]: Maybe this is a rebase problem? Should revert this change

File: plugins/octopus-deploy/src/api/index.ts [hunk 6]
```diff
@@ -106,13 +108,25 @@ export class OctopusDeployClient implements OctopusDeployApi {
     return responseJson;
   }
 
-  private async getApiUrl(projectId: string, releaseHistoryCount: number) {
+  private async getApiUrl(opts: {
+    projectId: string;
+    spaceId?: string;
```
Comment (freben) [inline plugins/octopus-deploy/src/api/index.ts:113]: Could have made sense to make these into `project: ProjectReference` instead, and make `ProjectReference` public (same above). But it's a matter of taste, feel free to ignore

File: plugins/octopus-deploy/README.md (hunk 4 above)
Comment (jmezach) [reply plugins/octopus-deploy/README.md:45]: Yeah, you're right. Not exactly sure what happened here, but I fixed the docs anyway.

File: plugins/octopus-deploy/package.json (hunk 5 above)
Comment (jmezach) [reply plugins/octopus-deploy/package.json:3]: Probably. Restored it.

File: plugins/octopus-deploy/src/api/index.ts (hunk 6 above)
Comment (jmezach) [reply plugins/octopus-deploy/src/api/index.ts]: I like that suggestion so I've made this change as well.

**Review by freben — APPROVED** [review]
Nice!

**Comment by JohnYoungers** [comment]
Thanks @jmezach ! We upgraded today and after a slight hiccup (I originally had the target as `https://octopus.domain.com` opposed to `https://octopus.domain.com/api`) I can confirm the change is working correctly with spaces.

Thanks again!

**Comment by jmezach** [comment]
@JohnYoungers Thanks for the feedback. Glad it worked out. We should probably improve the README a bit to explain that you need to include the `/api` postfix in the configured URL.
